File: bindings/node/src/zlink/contracts/messaging/subscription_event.ts
// SPDX-License-Identifier: MPL-2.0

import { RoutingId } from '../core/routing_id';

/** A subscriber's subscribe or unsubscribe as observed by an XPUB socket. */
export class SubscriptionEvent {
  /** The subscriber's routing id, or null when not known. */
  routingId: RoutingId | null;
  /** The topic that was subscribed or unsubscribed. */
  topic: string;
  /** True for a subscribe, false for an unsubscribe. */
  subscribed: boolean;

  /** Create an empty reusable event for use with `receiveSubscriptionEvent`. */
  constructor() {
    this.routingId = null;
    this.topic = '';
    this.subscribed = false;
  }
}

File: bindings/dotnet/src/Zlink/Contracts/Messaging/SubscriptionEvent.cs
// SPDX-License-Identifier: MPL-2.0

namespace Systems.Zlink;

/// <summary>
///     One active subscription: a topic filter and whether it is a pattern.
/// </summary>
/// <param name="Filter">The subscription topic or pattern.</param>
/// <param name="IsPattern">Whether <paramref name="Filter" /> is a pattern rather than an exact topic.</param>
public sealed record SubscriptionEntry(string Filter, bool IsPattern);

/// <summary>
///     A subscriber's subscribe or unsubscribe as observed by an XPUB socket.
/// </summary>
public sealed partial class SubscriptionEvent
{
    /// <summary>
    ///     Creates an empty event for reuse across receives.
    /// </summary>
    public SubscriptionEvent()
    {
    }

    /// <summary>
    ///     Gets the routing id of the subscriber, when known.
    /// </summary>
    public RoutingId? RoutingId { get; private set; }

    /// <summary>
    ///     Gets the topic that was subscribed or unsubscribed.
    /// </summary>
    public string Topic { get; private set; } = string.Empty;

    /// <summary>
    ///     Gets whether this event subscribed or unsubscribed the topic.
    /// </summary>
    public bool Subscribed { get; private set; }
}